lib/rspec/rails/matchers/active_job.rb

```diff
@@ -228,6 +234,38 @@ def does_not_match?(proc)
             !matches?(proc)
           end
         end
+
+        # @private
+        class HavePerformedJob < Base
+          FAILURE_MESSAGE_EXPECTATION_ACTION = 'perform'.freeze
+          MESSAGE_EXPECTATION_ACTION = 'performed'.freeze
+
+          def initialize(job)
+            super()
+            @job = job
+          end
+
+          def matches?(proc)
+            raise ArgumentError, "have_performed_job only supports block expectations" unless Proc === proc
+
+            original_performed_jobs_count = queue_adapter.performed_jobs.count
+            proc.call
+            in_block_jobs = queue_adapter.performed_jobs.drop(original_performed_jobs_count)
```

Comment:
Would it not be safer to use `-` here?
```
original_performed_jobs = queue_adapter.performed_jobs
proc.call
in_block_jobs = queue_adapter.performed_jobs - original_performed_jobs
```
I mean are jobs guaranteed to be ordered?

Reply:
I believe it's a copy-paste from [here](https://github.com/rspec/rspec-rails/blob/master/lib/rspec/rails/matchers/active_job.rb#L206).
It mostly worked like this (one old spec fails, one new):
```ruby
            original_performed_jobs = queue_adapter.performed_jobs.dup
            proc.call
            in_block_jobs = queue_adapter.performed_jobs - original_performed_jobs
```

However, since `drop` is non-destructive, it's more like a `tail`.

The culprit here is that if there were jobs of some type enqueued, and the block enqueues an identical job, `-` reports there was nothing enqueued, like in:
```ruby
[1, 1] - [1] # => []
```
With debug:
```
original_performed_jobs: {:job=>HeavyLiftingJob, :args=>[], :queue=>"default"}
queue_adapter.performed_jobs: {:job=>HeavyLiftingJob, :args=>[], :queue=>"default"}, {:job=>HeavyLiftingJob, :args=>[], :queue=>"default"}
    counts only jobs performed in block (FAILED - 1)

Failures:

  1) ActiveJob matchers have_performed_job counts only jobs performed in block
     Failure/Error:
       expect {
         heavy_lifting_job.perform_later
       }.to have_performed_job.exactly(1)

       expected to perform exactly 1 jobs, but performed 0
 ```

I really hope the ordering behaviour of `queue_adapter.performed_jobs` is the same as `queue_adapter.enqueued_jobs`, and we've relied on the latter for quite some time.

Reply:
Ok leave as is then.